### PR TITLE
Add support for using TransformerEngine

### PR DESCRIPTION
**Pull request by dskhudia**
Use TransformerEngine for either 1) linear layers only 2) Whole te.TransformerLayer to replace GPTBlock.

### CODE REVIEW

**Comment by abhi-mosaic** [comment]
Hi @dskhudia, can you comment on how necessary it is to use the `TransformerLayer` for performance / convergence?

For context, we have been making many changes to `GPTBlock` (and expect to make more variations), supporting different attention / arch / normalization strategies, and I'm not sure if we would end up using `TransformerLayer` much. The `te.Linear` approach feels more modular and we could use it with all the variations we end up building.

If it's still useful for testing purposes to have `TransformerLayer`, then I can think about a way to make the model construction more modular like `cfg.block_type=...`

File: examples/llm/main.py [hunk 1]
```diff
@@ -64,6 +69,23 @@ def main(cfg):
             "Reverting to `cfg.model.init_device='cpu'`.")
         cfg.model.init_device = 'cpu'
 
+    te_config = cfg.get('te_config', None)
+    te_config = om.to_container(te_config, resolve=True) if te_config else None
+    if not te_installed and te_config:
+        warnings.warn(
```
Comment (hanlint) [inline examples/llm/main.py:75]: Do we need to check whether it's an H100 and alert appropriately?

File: examples/llm/main.py [hunk 2]
```diff
@@ -64,6 +69,23 @@ def main(cfg):
...
+                "Transformer Engine is not installed but te_config is given! "
+                "Please install transformer engine with "
+                "pip install --upgrade git+https://github.com/NVIDIA/TransformerEngine.git@stable")
+    if not te_installed and not te_config['linear_only']:
+        print('use attn_impl == Triton to take care of key_padding_mask with TransformerEngine...')
```
Comment (hanlint) [inline examples/llm/main.py:80]: log instead of print. Message should also be more explicit: "Overriding attn_impl to use Triton to be compatible with Transformer Engine."

File: examples/llm/main.py [hunk 3]
```diff
@@ -64,6 +69,23 @@ def main(cfg):
...
+        cfg.attn_impl = 'triton'
+
+    if te_installed and te_config and init_device == 'meta':
+        warnings.warn(
+            "Using `cfg.model.init_device='meta'` doesn't work with transformer engine! "
```
Comment (hanlint) [inline examples/llm/main.py:85]:
```suggestion
            "Using `cfg.model.init_device='meta'` not compatible with Transformer Engine."
```

File: examples/llm/main.py [hunk 4]
```diff
@@ -64,6 +69,23 @@ def main(cfg):
...
+            "Reverting to `cfg.model.init_device='cpu'`.")
+        cfg.model.init_device = 'cpu'
```
Comment (hanlint) [inline examples/llm/main.py:87]: open question: should we be overriding settings vs. throwing an error? e.g. forcing the init device out of meta can have other implications, and maybe forcing the user to switch it might be a better experience?

File: examples/llm/src/mosaic_gpt.py [hunk 5]
```diff
@@ -205,13 +210,21 @@ class GPTMLP(nn.Module):
 
     def __init__(self, cfg: DictConfig, device: Optional[str] = None):
         super().__init__()
-        self.mlp_up = nn.Linear(cfg.d_model,
-                                cfg.mlp_ratio * cfg.d_model,
-                                device=device)
+        if te_installed and cfg.get('te_config.linear_only'):
```
Comment (hanlint) [inline examples/llm/src/mosaic_gpt.py:213]: should the requisite warnings (e.g. if te_config provided but te not installed) but put here as well, if users are creating their own models and not using `main.py` ?

**Comment by SinanAkkoyun** [comment]
Hello!
How does one use this with inference of the model? We have a great interest in that.

Best Regards

**Comment by dakinggg** [comment]
Closing, because llm has moved to llm-foundry.
